roachtest: scrub/all-checks/tpcc/w=1000 failed

SHA: https://github.com/cockroachdb/cockroach/commits/3a7ea2d8c9d4a3e0d97f8f106fcf95b3f03765ec

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc/w=1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1187480&tab=buildLog

```
The test failed on master:
	cluster.go:1267,tpcc.go:130,cluster.go:1605,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1187480-scrub-all-checks-tpcc-w-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		9.2 103079.2 newOrder
		   59m6s    26598            0.0            2.1      0.0      0.0      0.0      0.0 orderStatus
		   59m6s    26598            0.0           20.0      0.0      0.0      0.0      0.0 payment
		   59m6s    26598            0.0            2.1      0.0      0.0      0.0      0.0 stockLevel
		E190320 15:17:10.469844 1 workload/cli/run.go:420  error in payment: dial tcp 10.142.0.62:26257: connect: connection refused
		   59m7s    71358            0.0            2.2      0.0      0.0      0.0      0.0 delivery
		   59m7s    71358            1.0           21.1 103079.2 103079.2 103079.2 103079.2 newOrder
		   59m7s    71358            0.0            2.1      0.0      0.0      0.0      0.0 orderStatus
		   59m7s    71358            0.0           20.0      0.0      0.0      0.0      0.0 payment
		   59m7s    71358            0.0            2.1      0.0      0.0      0.0      0.0 stockLevel
		E190320 15:17:11.469848 1 workload/cli/run.go:420  error in payment: dial tcp 10.142.0.62:26257: connect: connection refused
		: signal: killed
	cluster.go:1626,tpcc.go:140,scrub.go:58,test.go:1214: unexpected node event: 3: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/9399d559ae196e5cf2ad122195048ff9115ab56a

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc/w=1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1194326&tab=buildLog

```
The test failed on release-19.1:
	scrub.go:83,cluster.go:1605,errgroup.go:57: read tcp 172.17.0.2:60850->35.237.203.215:26257: read: connection reset by peer
	cluster.go:1267,tpcc.go:130,cluster.go:1605,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1194326-scrub-all-checks-tpcc-w-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		.0            0.0      0.0      0.0      0.0      0.0 newOrder
		   4m36s    22663            0.0            0.0      0.0      0.0      0.0      0.0 orderStatus
		   4m36s    22663            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   4m36s    22663            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		E190323 14:51:44.563428 1 workload/cli/run.go:420  error in payment: dial tcp 10.142.0.20:26257: connect: connection refused
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		   4m37s    86325            0.0            0.0      0.0      0.0      0.0      0.0 delivery
		   4m37s    86325            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		   4m37s    86325            0.0            0.0      0.0      0.0      0.0      0.0 orderStatus
		   4m37s    86325            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   4m37s    86325            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1626,tpcc.go:140,scrub.go:58,test.go:1214: Goexit() was called

```

SHA: https://github.com/cockroachdb/cockroach/commits/a039a93a5cc6eb3f395ceb6f7dc8030985dccc29

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc/w=1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1212269&tab=buildLog

```
The test failed on release-19.1:
	cluster.go:1293,tpcc.go:130,cluster.go:1631,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1212269-scrub-all-checks-tpcc-w-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		0            0.0      0.0      0.0      0.0      0.0 newOrder
		    5m0s    74395            0.0            0.0      0.0      0.0      0.0      0.0 orderStatus
		    5m0s    74395            0.0            0.0      0.0      0.0      0.0      0.0 payment
		    5m0s    74395            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		E190401 14:40:54.130813 1 workload/cli/run.go:420  error in payment: dial tcp 10.142.0.139:26257: connect: connection refused
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		    5m1s    87634            0.0            0.0      0.0      0.0      0.0      0.0 delivery
		    5m1s    87634            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		    5m1s    87634            0.0            0.0      0.0      0.0      0.0      0.0 orderStatus
		    5m1s    87634            0.0            0.0      0.0      0.0      0.0      0.0 payment
		    5m1s    87634            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1652,tpcc.go:140,scrub.go:58,test.go:1223: unexpected node event: 2: dead
	cluster.go:953,context.go:90,cluster.go:942,asm_amd64.s:522,panic.go:397,test.go:774,test.go:760,cluster.go:1652,tpcc.go:140,scrub.go:58,test.go:1223: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1212269-scrub-all-checks-tpcc-w-1000 --oneshot --ignore-empty-nodes: exit status 1 5: skipped
		2: dead
		1: 3931
		4: 3311
		3: 3253
		Error:  2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/c6df752eefe4609b8a5bbada0955f79a2cfb790e

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc/w=1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1217763&tab=buildLog

```
The test failed on master:
	scrub.go:83,cluster.go:1667,errgroup.go:57: dial tcp 35.237.106.233:26257: connect: connection refused
	cluster.go:1329,tpcc.go:130,cluster.go:1667,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1217763-scrub-all-checks-tpcc-w-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		.0      0.0 newOrder
		  21m42s    34983            0.0            1.0      0.0      0.0      0.0      0.0 orderStatus
		  21m42s    34983            0.0            8.6      0.0      0.0      0.0      0.0 payment
		  21m42s    34983            0.0            0.9      0.0      0.0      0.0      0.0 stockLevel
		E190403 14:28:39.898324 1 workload/cli/run.go:420  error in payment: dial tcp 10.142.0.62:26257: connect: connection refused
		  21m43s    72479            0.0            1.1      0.0      0.0      0.0      0.0 delivery
		  21m43s    72479            0.0            9.5      0.0      0.0      0.0      0.0 newOrder
		  21m43s    72479            0.0            1.0      0.0      0.0      0.0      0.0 orderStatus
		  21m43s    72479            0.0            8.6      0.0      0.0      0.0      0.0 payment
		  21m43s    72479            0.0            0.9      0.0      0.0      0.0      0.0 stockLevel
		E190403 14:28:40.898314 1 workload/cli/run.go:420  error in newOrder: dial tcp 10.142.0.62:26257: connect: connection refused
		: signal: killed
	cluster.go:1688,tpcc.go:140,scrub.go:58,test.go:1228: unexpected node event: 1: dead
	cluster.go:953,context.go:90,cluster.go:942,asm_amd64.s:522,panic.go:397,test.go:776,test.go:762,cluster.go:1688,tpcc.go:140,scrub.go:58,test.go:1228: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1217763-scrub-all-checks-tpcc-w-1000 --oneshot --ignore-empty-nodes: exit status 1 5: skipped
		1: dead
		3: 4074
		4: 3907
		2: 3701
		Error:  1: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/b88a6ce86bfe507e14e1e80fdaefd219b5f0b046

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc/w=1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1222804&tab=buildLog

```
The test failed on provisional_201904050025_v19.1.0-rc.2:
	cluster.go:1329,tpcc.go:130,cluster.go:1667,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1222804-scrub-all-checks-tpcc-w-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		0      0.0 newOrder
		   7m26s    45579            0.0            0.1      0.0      0.0      0.0      0.0 orderStatus
		   7m26s    45579            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   7m26s    45579            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		E190405 14:00:53.004702 1 workload/cli/run.go:420  error in newOrder: dial tcp 10.128.0.10:26257: connect: connection refused
		   7m27s   117069            0.0            0.0      0.0      0.0      0.0      0.0 delivery
		   7m27s   117069            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		   7m27s   117069            0.0            0.1      0.0      0.0      0.0      0.0 orderStatus
		   7m27s   117069            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   7m27s   117069            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		E190405 14:00:54.004722 1 workload/cli/run.go:420  error in newOrder: dial tcp 10.128.0.10:26257: connect: connection refused
		: signal: killed
	cluster.go:1688,tpcc.go:140,scrub.go:58,test.go:1228: unexpected node event: 4: dead
	cluster.go:953,context.go:90,cluster.go:942,asm_amd64.s:522,panic.go:397,test.go:776,test.go:762,cluster.go:1688,tpcc.go:140,scrub.go:58,test.go:1228: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1222804-scrub-all-checks-tpcc-w-1000 --oneshot --ignore-empty-nodes: exit status 1 5: skipped
		4: dead
		1: 3644
		3: 3288
		2: 3411
		Error:  4: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/6da68d7fe2c9a29b85e2ec0c7e545a0d6bdc4c5c

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc/w=1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1226521&tab=buildLog

```
The test failed on release-19.1:
	scrub.go:83,cluster.go:1667,errgroup.go:57: read tcp 172.17.0.2:39498->34.73.1.69:26257: read: connection reset by peer
	cluster.go:1329,tpcc.go:169,cluster.go:1667,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1226521-scrub-all-checks-tpcc-w-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		      0.0 newOrder
		  1h1m2s    25858            1.0            2.2   4563.4   4563.4   4563.4   4563.4 orderStatus
		  1h1m2s    25858            2.0           20.4  47244.6 103079.2 103079.2 103079.2 payment
		  1h1m2s    25858            0.0            2.1      0.0      0.0      0.0      0.0 stockLevel
		E190407 15:10:28.426806 1 workload/cli/run.go:420  error in delivery: dial tcp 10.142.0.152:26257: connect: connection refused
		  1h1m3s    78871            0.0            2.2      0.0      0.0      0.0      0.0 delivery
		  1h1m3s    78871            0.0           21.6      0.0      0.0      0.0      0.0 newOrder
		  1h1m3s    78871            0.0            2.2      0.0      0.0      0.0      0.0 orderStatus
		  1h1m3s    78871            0.0           20.4      0.0      0.0      0.0      0.0 payment
		  1h1m3s    78871            0.0            2.1      0.0      0.0      0.0      0.0 stockLevel
		E190407 15:10:29.426873 1 workload/cli/run.go:420  error in payment: dial tcp 10.142.0.152:26257: connect: connection refused
		: signal: killed
	cluster.go:1688,tpcc.go:179,scrub.go:58,test.go:1237: Goexit() was called
	cluster.go:953,context.go:90,cluster.go:942,asm_amd64.s:522,panic.go:397,test.go:785,test.go:771,cluster.go:1688,tpcc.go:179,scrub.go:58,test.go:1237: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1226521-scrub-all-checks-tpcc-w-1000 --oneshot --ignore-empty-nodes: exit status 1 5: skipped
		1: dead
		2: 3588
		4: 3560
		3: 3699
		Error:  1: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/682c2f2f466bbf768545ca4687822206a63983ad

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc/w=1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1231772&tab=buildLog

```
The test failed on master:
	cluster.go:1329,tpcc.go:169,cluster.go:1667,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1231772-scrub-all-checks-tpcc-w-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		un.go:428  error in payment: dial tcp 10.142.0.174:26257: connect: connection refused
		   2m18s     8423            0.0            0.0      0.0      0.0      0.0      0.0 delivery
		   2m18s     8423            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		   2m18s     8423            0.0            0.0      0.0      0.0      0.0      0.0 orderStatus
		   2m18s     8423            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   2m18s     8423            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		   2m19s     8423            0.0            0.0      0.0      0.0      0.0      0.0 delivery
		   2m19s     8423            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		   2m19s     8423            0.0            0.0      0.0      0.0      0.0      0.0 orderStatus
		   2m19s     8423            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   2m19s     8423            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1405,cluster.go:1424,cluster.go:1528,scrub.go:72,cluster.go:1667,errgroup.go:57: context canceled
	cluster.go:1688,tpcc.go:179,scrub.go:58,test.go:1237: unexpected node event: 2: dead
	cluster.go:953,context.go:90,cluster.go:942,asm_amd64.s:522,panic.go:397,test.go:785,test.go:771,cluster.go:1688,tpcc.go:179,scrub.go:58,test.go:1237: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1231772-scrub-all-checks-tpcc-w-1000 --oneshot --ignore-empty-nodes: exit status 1 5: skipped
		2: dead
		1: 3244
		3: 3057
		4: 3061
		Error:  2: dead

```

I repro'd a check failure last night: 

```
--- FAIL: scrub/all-checks/tpcc/w=1000 2551.30s
	cluster.go:1468,tpcc.go:146,scrub.go:58,test.go:1841: /Users/bdarnell/cockroach/bin/roachprod run ben-s-int64-1554937051-08-n5cpu4:5 -- ./workload check tpcc --warehouses=1000 {pgurl:1} returned:
		stderr:
		
		stdout:
		I190411 04:49:00.217858 1 workload/tpcc/tpcc.go:292  check 3.3.2.1 took 504.910721ms
		Error: check failed: 3.3.2.1: 1 rows returned, expected zero
		Error:  exit status 1
		: exit status 1
```

It's a different check this time; this one is simpler and only involves a single query. And it's persistent. I can reproduce the failure and see the data that it reports as inconsistent:

```
root@:26257/tpcc> SELECT *                                                                                                                                                     FROM warehouse                                                                                                                                                                 FULL OUTER JOIN                                                                                                                                                                (SELECT d_w_id, sum(d_ytd) as sum_d_ytd FROM district GROUP BY d_w_id)                                                                                                         ON (w_id = d_w_id)                                                                                                                                                             WHERE w_ytd != sum_d_ytd;
  w_id | w_name | w_street_1 | w_street_2 | w_city | w_state |   w_zip   | w_tax  |   w_ytd   | d_w_id | sum_d_ytd  
+------+--------+------------+------------+--------+---------+-----------+--------+-----------+--------+-----------+
   626 | 10     | 18         | 16         | 15     | ER      | 897611111 | 0.0638 | 392524.52 |    626 | 395535.05  
(1 row)

Time: 65.283362ms
```

SHA: https://github.com/cockroachdb/cockroach/commits/9e6ae3cc37e7691147bb6f5d1a156ebe4c5cf7f9

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc/w=1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1245443&tab=buildLog

```
The test failed on master:
	cluster.go:1329,tpcc.go:168,cluster.go:1667,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1245443-scrub-all-checks-tpcc-w-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		n.go:428  error in newOrder: dial tcp 10.142.0.111:26257: connect: connection refused
		   9m11s    20272            0.0            0.0      0.0      0.0      0.0      0.0 delivery
		   9m11s    20272            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		   9m11s    20272            0.0            0.4      0.0      0.0      0.0      0.0 orderStatus
		   9m11s    20272            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   9m11s    20272            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		   9m12s    20272            0.0            0.0      0.0      0.0      0.0      0.0 delivery
		   9m12s    20272            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		   9m12s    20272            0.0            0.4      0.0      0.0      0.0      0.0 orderStatus
		   9m12s    20272            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   9m12s    20272            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1688,tpcc.go:178,scrub.go:58,test.go:1237: unexpected node event: 4: dead
	cluster.go:953,context.go:89,cluster.go:942,asm_amd64.s:522,panic.go:397,test.go:785,test.go:771,cluster.go:1688,tpcc.go:178,scrub.go:58,test.go:1237: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1245443-scrub-all-checks-tpcc-w-1000 --oneshot --ignore-empty-nodes: exit status 1 5: skipped
		4: dead
		2: 3395
		1: 3226
		3: 3295
		Error:  4: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/83de585d331b05a4aa02a65b353bed6bf829b696

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc/w=1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1247383&tab=buildLog

```
The test failed on master:
	cluster.go:1329,tpcc.go:168,cluster.go:1667,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1247383-scrub-all-checks-tpcc-w-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		n.go:428  error in newOrder: dial tcp 10.142.0.111:26257: connect: connection refused
		   3m14s    14732            0.0            0.0      0.0      0.0      0.0      0.0 delivery
		   3m14s    14732            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		   3m14s    14732            0.0            0.0      0.0      0.0      0.0      0.0 orderStatus
		   3m14s    14732            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   3m14s    14732            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		   3m15s    14754            0.0            0.0      0.0      0.0      0.0      0.0 delivery
		   3m15s    14754            0.0            0.0      0.0      0.0      0.0      0.0 newOrder
		   3m15s    14754            0.0            0.0      0.0      0.0      0.0      0.0 orderStatus
		   3m15s    14754            0.0            0.0      0.0      0.0      0.0      0.0 payment
		   3m15s    14754            0.0            0.0      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	cluster.go:1405,cluster.go:1424,cluster.go:1528,scrub.go:72,cluster.go:1667,errgroup.go:57: context canceled
	cluster.go:1688,tpcc.go:178,scrub.go:58,test.go:1237: unexpected node event: 4: dead
	cluster.go:953,context.go:89,cluster.go:942,asm_amd64.s:522,panic.go:397,test.go:785,test.go:771,cluster.go:1688,tpcc.go:178,scrub.go:58,test.go:1237: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1247383-scrub-all-checks-tpcc-w-1000 --oneshot --ignore-empty-nodes: exit status 1 5: skipped
		4: dead
		3: 4022
		1: 3969
		2: 4188
		Error:  4: dead
	test.go:1225: test timed out (11h49m3.093261994s)
	test.go:995,asm_amd64.s:523,panic.go:513,log.go:219,cluster.go:1020,context.go:89,cluster.go:1008,test.go:1191,asm_amd64.s:522,panic.go:397,test.go:785,test.go:775,cluster.go:953,context.go:89,cluster.go:942,asm_amd64.s:522,panic.go:397,test.go:785,test.go:771,cluster.go:1688,tpcc.go:178,scrub.go:58,test.go:1237: write /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190418-1247383/scrub/all-checks/tpcc/w=1000/test.log: file already closed

```

--- a/pkg/storage/node_liveness.go
+++ b/pkg/storage/node_liveness.go
@@ -53,6 +53,11 @@ var (
 	// the underlying liveness record has had its epoch incremented.
 	ErrEpochIncremented = errors.New("heartbeat failed on epoch increment")
 
+	// ErrEpochAlreadyIncremented is returned by IncrementEpoch when
+	// someone else has already incremented the epoch to the desired
+	// value.
+	ErrEpochAlreadyIncremented = errors.New("epoch already incremented")
+
 	errLiveClockNotLive = errors.New("not live")
 )
 
@@ -685,8 +690,6 @@ func (nl *NodeLiveness) getLivenessLocked(nodeID roachpb.NodeID) (*storagepb.Liv
 	return nil, ErrNoLivenessRecord
 }
 
-var errEpochAlreadyIncremented = errors.New("epoch already incremented")
-
 // IncrementEpoch is called to increment the current liveness epoch,
 // thereby invalidating anything relying on the liveness of the
 // previous epoch. This method does a conditional put on the node
@@ -714,15 +717,12 @@ func (nl *NodeLiveness) IncrementEpoch(ctx context.Context, liveness *storagepb.
 	if err := nl.updateLiveness(ctx, update, liveness, func(actual storagepb.Liveness) error {
 		defer nl.maybeUpdate(actual)
 		if actual.Epoch > liveness.Epoch {
-			return errEpochAlreadyIncremented
+			return ErrEpochAlreadyIncremented
 		} else if actual.Epoch < liveness.Epoch {
 			return errors.Errorf("unexpected liveness epoch %d; expected >= %d", actual.Epoch, liveness.Epoch)
 		}
 		return errors.Errorf("mismatch incrementing epoch for %+v; actual is %+v", *liveness, actual)
 	}); err != nil {
-		if err == errEpochAlreadyIncremented {
-			return nil
-		}
 		return err
 	}
 

--- a/pkg/storage/node_liveness_test.go
+++ b/pkg/storage/node_liveness_test.go
@@ -316,8 +316,8 @@ func TestNodeLivenessEpochIncrement(t *testing.T) {
 		t.Errorf("expected epoch increment == 1; got %d", c)
 	}
 
-	// Verify noop on incrementing an already-incremented epoch.
-	if err := mtc.nodeLivenesses[0].IncrementEpoch(context.Background(), oldLiveness); err != nil {
+	// Verify error on incrementing an already-incremented epoch.
+	if err := mtc.nodeLivenesses[0].IncrementEpoch(context.Background(), oldLiveness); err != storage.ErrEpochAlreadyIncremented {
 		t.Fatalf("unexpected error incrementing a non-live node: %s", err)
 	}
 
@@ -610,7 +610,7 @@ func TestNodeLivenessConcurrentIncrementEpochs(t *testing.T) {
 		}()
 	}
 	for i := 0; i < concurrency; i++ {
-		if err := <-errCh; err != nil {
+		if err := <-errCh; err != nil && err != storage.ErrEpochAlreadyIncremented {
 			t.Fatalf("concurrent increment epoch %d failed: %s", i, err)
 		}
 	}

--- a/pkg/storage/replica_range_lease.go
+++ b/pkg/storage/replica_range_lease.go
@@ -303,7 +303,38 @@ func (p *pendingLeaseRequest) requestLeaseAsync(
 							log.Info(ctx, err)
 						}
 					} else if err = p.repl.store.cfg.NodeLiveness.IncrementEpoch(ctx, status.Liveness); err != nil {
-						log.Error(ctx, err)
+						// If we get ErrEpochAlreadyIncremented, someone else beat
+						// us to it. This proves that the target node is truly
+						// dead *now*, but it doesn't prove that it was dead at
+						// status.Timestamp (which we've encoded into our lease
+						// request). It's possible that the node was temporarily
+						// considered dead but revived without having its epoch
+						// incremented, i.e. that it was in fact live at
+						// status.Timestamp.
+						//
+						// It would be incorrect to simply proceed to sending our
+						// lease request since our lease.Start may precede the
+						// effective end timestamp of the predecessor lease (the
+						// expiration of the last successful heartbeat before the
+						// epoch increment), and so under this lease this node's
+						// timestamp cache would not necessarily reflect all reads
+						// served by the prior leaseholder.
+						//
+						// It would be correct to bump the timestamp in the lease
+						// request and proceed, but that just sets up another race
+						// between this node and the one that already incremented
+						// the epoch. They're probably going to beat us this time
+						// too, so just return the NotLeaseHolderError here
+						// instead of trying to fix up the timestamps and submit
+						// the lease request.
+						//
+						// ErrEpochAlreadyIncremented is not an unusual situation,
+						// so we don't log it as an error.
+						//
+						// https://github.com/cockroachdb/cockroach/issues/35986
+						if err != ErrEpochAlreadyIncremented {
+							log.Error(ctx, err)
+						}
 					}
 				}
 				// Set error for propagation to all waiters below.

